[DOCUMENTATION][feature/otx] Explanation section. PR 2 (#1566)
open-edge-platform/training_extensions: docs/source/explanation/additional_features/auto_configuration.rst

Auto-configuration
==================

> Auto configuration for a deep learning framework refers to the process of automatically determining the most appropriate settings for the framework's parameters, such as specific algorithm and set of hyperparameters, based on the dataset and the specific task at hand.

Reviewer: ```suggestion
Auto-configuration for a deep learning framework means the automatic finding of the most appropriate settings for the training parameters, based on the dataset and the specific task at hand.
```

> Auto configuration can help to save time, ease the process of interaction with OTX CLI and give a better baseline for the given dataset.

Reviewer: ```suggestion
Auto-configuration can help to save time, ease the process of interaction with OTX CLI and give a better baseline for the given dataset.
```

At this end, we developed a simple auto-configuration functionality to ease the process of training and validation utilizing our framework.

> Basically, to start training and obtain a good baseline with best trade off between accuracy and speed we need to pass only a dataset in the right format

Reviewer: ```suggestion
Basically, to start the training and obtain a good baseline with the best trade-off between accuracy and speed we need to pass only a dataset in the right format
```

without specifiing anything else:

.. code-block::

    $ otx train --train-data-root <path_to_data_root>

> Moreover, our dataset can have no train/val splits at all. The Datumaro manager integrated to OTX will handle it on its own.

Reviewer: ```suggestion
Moreover, our dataset can have no train/val splits at all. The Datumaro manager integrated into OTX will handle it on its own.
```

It will recognize the task by analizing the dataset and if there is no splits for the validation - Datumaro will do random auto-split, memorizing this split to further use it with ``otx eval`` and ``otx optimize``.

> Currently, Datumaro auto-split feature supports 3 tasks: multi-class classification, detection and semantic segmentation.

Reviewer: Can you highlight task names using bold or italic font, please?

For now, after dataset preparation the training will be started with the middle-sized template to achieve a competitive accuracy preserving fast inference.

Supported dataset formats for each task:

    - classificaiton: Imagenet, COCO (multi-label), :ref:`custom hierarchical <hierarchical_dataset>`
    - object detection: COCO, Pascal-VOC, YOLO
    - semantic segmentation: common semantic segmentation, Pascal-VOC, cityscapes, ADE20k
    - action classification: CVAT
    - action : CVAT
    - anomaly classificaiton: MVTec
    - anomaly detection: MVTec
    - anomaly segmentation: MVTec

>     - instance segmentation: COCO, Pascal-VOC

Reviewer: I'd suggest removing indents here, because it's rendered similarly to Note field
![image](https://user-images.githubusercontent.com/33455576/214813396-f89031c8-cf09-4ce2-bec1-1ce264fc7ddb.png)

Reviewer: I would also add links to these formats, especially because you already have them in Segmentation section `COCO <https://cocodataset.org/#format-data>` and `Pascal-VOC <https://openvinotoolkit.github.io/datumaro/docs/formats/pascal_voc/>`_

> If we have a dataset format that occluded with other tasks, for example ``COCO`` format, we should directly emphasize the task type with additional CLI option:

Reviewer: ```suggestion
If we have a dataset format that occluded with other tasks, for example ``COCO`` format, we should directly emphasize the task type with an additional CLI option:
```

Reviewer: Also, I'd personally put this into Note field

Auto-configuration
==================

Auto-configuration for a deep learning framework means the automatic finding of the most appropriate settings for the training parameters, based on the dataset and the specific task at hand.
Auto-configuration can help to save time, ease the process of interaction with OTX CLI and give a better baseline for the given dataset.

At this end, we developed a simple auto-configuration functionality to ease the process of training and validation utilizing our framework.
Basically, to start the training and obtain a good baseline with the best trade-off between accuracy and speed we need to pass only a dataset in the right format without specifiing anything else:

.. code-block::

> 

Reviewer 2: Auto-configuration also supports `otx build --train-data-roots <path_to_data_root>`

Author: Is there a reason to run `otx build` instead of `otx train` for users? If we call `otx build` it means `otx` will create a folder with all components, but the template (model) will be the middle one. So, using auto-config calling otx build is unnecessary, IMO. But, I will add it as a note mentioning that `otx build` can be called to create a folder.

    $ otx train --train-data-root <path_to_data_root>

Moreover, our dataset can have no train/val splits at all. The Datumaro manager integrated into OTX will handle it on its own.
It will recognize the task by analizing the dataset and if there is no splits for the validation - Datumaro will do random auto-split, memorizing this split to further use it with ``otx eval`` and ``otx optimize``.
Currently, Datumaro auto-split feature supports 3 tasks: **multi-class classification**, **detection** and **semantic segmentation**.
After dataset preparation the training will be started with the middle-sized template to achieve a competitive accuracy preserving fast inference.

> Supported dataset formats for each task:

Reviewer 2: Since it is a description for auto-configuration, how about removing the datasets that can't be used for auto-configurations?

For this, I need to investigate time to analyze which dataset can be used for auto-configuration. Currently, Only 3 datasets (`Imagenet (cls)` `COCO (det)` `Cityscape (seg)`) can be auto-splitted by Datumaro

Author: If we have splits for other datasets and task can we use a simple auto-config?

Author: I mean running `otx train --train-data-roots <path_to_data_root>` with other formats if we have validation folder exactly with the same name as it should be in original dataset format

Reviewer 2: If we don't use auto-split function, then we can use datasets below:
Classification: ImageNet
Detection: COCO, VOC, YOLO
Segmentation: Common Semantic Segmentation, VOC, Cityscape, ADE20K 2017, 2020

`otx train --train-data-roots <coco_path>`

Author: I moved supported dataset formats into note section. It is a good idea to give an overview about all supported datasets. We also have a note about supported formats by auto-split

- semantic segmentation: `Common Semantic Segmentation <https://openvinotoolkit.github.io/datumaro/docs/formats/common_semantic_segmentation/>`_, `Pascal-VOC <https://openvinotoolkit.github.io/datumaro/docs/formats/pascal_voc/>`_, `Cityscapes <https://openvinotoolkit.github.io/datumaro/docs/formats/cityscapes/>`_, `ADE20k <https://openvinotoolkit.github.io/datumaro/docs/formats/ade20k2020/>`_
- action classification: `CVAT <https://opencv.github.io/cvat/docs/manual/advanced/xml_format/>`_
- action : `CVAT <https://opencv.github.io/cvat/docs/manual/advanced/xml_format/>`_
- anomaly classificaiton: `MVTec <https://www.mvtec.com/company/research/datasets/mvtec-ad>`_
- anomaly detection: `MVTec <https://www.mvtec.com/company/research/datasets/mvtec-ad>`_
- anomaly segmentation: `MVTec <https://www.mvtec.com/company/research/datasets/mvtec-ad>`_
- instance segmentation: `COCO <https://cocodataset.org/#format-data>`_, `Pascal-VOC <https://openvinotoolkit.github.io/datumaro/docs/formats/pascal_voc/>`_

.. note::

>     If we have a dataset format that occluded with other tasks, for example ``COCO`` format, we should directly emphasize the task type with an additional CLI option:

Reviewer 2: Want to add some contents

[Before]
If we have a dataset format that occluded with other tasks, for example ``COCO`` format, we should directly emphasize the task type with an additional CLI option:

[After]
If we have a dataset format that occluded with other tasks, for example ``COCO`` format, we should directly emphasize the task type with an additional CLI option. If not, OTX automatically chooses the task type that you might doesn't intend:

Auto-configuration
==================

Auto-configuration for a deep learning framework means the automatic finding of the most appropriate settings for the training parameters, based on the dataset and the specific task at hand.

> Auto-configuration can help to save time, ease the process of interaction with OTX CLI and give a better baseline for the given dataset.

Reviewer: ```suggestion
Auto-configuration can help to save time, it eases the process of interaction with OTX CLI and gives a better baseline for the given dataset.
```

At this end, we developed a simple auto-configuration functionality to ease the process of training and validation utilizing our framework.
Basically, to start the training and obtain a good baseline with the best trade-off between accuracy and speed we need to pass only a dataset in the right format without specifiing anything else:

.. code-block::

    $ otx train --train-data-root <path_to_data_root>

.. note::

    OTX supports also ``otx build`` mode with the auto-configuration feature. We can build OTX workspace with the following CLI command:

    .. code-block::

        $ otx build --train-data-root <path_to_data_root>

Moreover, our dataset can have no train/val splits at all. The Datumaro manager integrated into OTX will handle it on its own.

> It will recognize the task by analizing the dataset and if there is no splits for the validation - Datumaro will do random auto-split, memorizing this split to further use it with ``otx eval`` and ``otx optimize``.

Reviewer 3: ```suggestion
It will recognize the task by analyzing the dataset and if there is no splits for the validation - Datumaro will do random auto-split, memorizing this split to further use it with ``otx eval`` and ``otx optimize``.
```

.. note::

>     Currently, Datumaro auto-split feature supports 3 formats: `Imagenet <https://www.image-net.org/>`_  (**multi-class classification**), `COCO <https://cocodataset.org/#format-data>`_ (**detection**) and `Cityscapes <https://openvinotoolkit.github.io/datumaro/docs/formats/cityscapes/>`_ (**semantic segmentation**).

Reviewer: I'd remove highlighting tasks (only here) because links and bold font together look overwhelming.

    - action classification: `CVAT <https://opencv.github.io/cvat/docs/manual/advanced/xml_format/>`_
    - action : `CVAT <https://opencv.github.io/cvat/docs/manual/advanced/xml_format/>`_
    - anomaly classificaiton: `MVTec <https://www.mvtec.com/company/research/datasets/mvtec-ad>`_
    - anomaly detection: `MVTec <https://www.mvtec.com/company/research/datasets/mvtec-ad>`_
    - anomaly segmentation: `MVTec <https://www.mvtec.com/company/research/datasets/mvtec-ad>`_
    - instance segmentation: `COCO <https://cocodataset.org/#format-data>`_, `Pascal-VOC <https://openvinotoolkit.github.io/datumaro/docs/formats/pascal_voc/>`_

    If we have a dataset format occluded with other tasks, for example ``COCO`` format, we should directly emphasize the task type with an additional CLI option. If not, OTX automatically chooses the task type that you might don't intend:

    .. code-block::

>         $ otx train --train-data-root <path_to_data_root> --task {CLASSIFICATION, DETECTION, SEGMENTATION, ACTION_CLASSIFICATION, ACTION_DETECTION, ANOMALY_CLASSIFICATION, ANOMALY_DETECTION, ANOMALY_SEGMENTATION, INSTANCE_SEGMENTATION}

Reviewer 2: Current `build` and `train` CLI have different arguments,

For the `otx build` --> `--model`, `--backbone`, `--train_type`, `--task`, ...
For the `otx train` --> `--train/val-data-roots`, `--template`, `--save-model-to`, ...

`otx build --train-data-root {}` --> OK.
`otx build --train-data-root {} --task {}` --> OK.
`otx build --train-data-root {} --template {}` --> Can't

`otx train --train-data-root {}` --> OK.
`otx train --train-data-root {} --task {}` --> Can't
`otx train --train-data-root {} --template {}` --> OK.

Author: I changed `train` to `build` in case when we have occluded datasets

Auto-configuration
==================

Auto-configuration for a deep learning framework means the automatic finding of the most appropriate settings for the training parameters, based on the dataset and the specific task at hand.
Auto-configuration can help to save time, it eases the process of interaction with OTX CLI and gives a better baseline for the given dataset.

At this end, we developed a simple auto-configuration functionality to ease the process of training and validation utilizing our framework.
Basically, to start the training and obtain a good baseline with the best trade-off between accuracy and speed we need to pass only a dataset in the right format without specifying anything else:

.. code-block::

    $ otx train --train-data-root <path_to_data_root>

.. note::

    OTX supports also ``otx build`` mode with the auto-configuration feature. We can build OTX workspace with the following CLI command:

    .. code-block::

        $ otx build --train-data-root <path_to_data_root>

Moreover, our dataset can have no train/val splits at all. The Datumaro manager integrated into OTX will handle it on its own.

> It will recognize the task by analyzing the dataset and if there is no splits for the validation - Datumaro will do a random auto-split, memorizing this split to further use it with ``otx eval`` and ``otx optimize``.

Reviewer 2: One small suggestion, 

Since `otx build` will makes the `train` and `val` datasets, users can't directly use these datasets to `otx eval` which is required `test` dataset. In the case of `otx optimize`,there is no problem. 

So, how about converting like below?

It will recognize the task by analyzing the dataset and if there is no splits for the validation - Datumaro will do a random auto-split, saving this split to the workspace. It could be used with ``otx optimize`` or ``otx train``.

Author: Okay, maybe this is better, I changed it.
